pyiceberg/table/__init__.py

```diff
@@ -2271,3 +2331,242 @@ def commit(self) -> Snapshot:
+                        self._table.schema(), field.source_id, field.field_id, renamed, field.transform, partition_names
+                    )
+                else:
+                    new_field = _add_new_field(
+                        self._table.schema(), field.source_id, field.field_id, field.name, field.transform, partition_names
+                    )
+                partition_fields.append(new_field)
+            elif self._table.format_version == 1:
+                renamed = self._renames.get(field.name)
+                if renamed:
+                    new_field = _add_new_field(
+                        self._table.schema(), field.source_id, field.field_id, renamed, VoidTransform(), partition_names
+                    )
+                else:
+                    new_field = _add_new_field(
+                        self._table.schema(), field.source_id, field.field_id, field.name, VoidTransform(), partition_names
+                    )
+
+                partition_fields.append(new_field)
+
+        for added_field in self._adds:
+            new_field = PartitionField(
+                source_id=added_field.source_id,
+                field_id=added_field.field_id,
+                transform=added_field.transform,
+                name=added_field.name,
+            )
+            partition_fields.append(new_field)
+
+        # Reuse spec id or create a new one.
+        new_spec = PartitionSpec(*partition_fields)
+        new_spec_id = INITIAL_PARTITION_SPEC_ID
+        for spec in self._table.specs().values():
+            if new_spec.compatible_with(spec):
+                new_spec_id = spec.spec_id
+                break
+            elif new_spec_id <= spec.spec_id:
+                new_spec_id = spec.spec_id + 1
+        return PartitionSpec(*partition_fields, spec_id=new_spec_id)
+
+    def _redundant_time_partition(self, field: PartitionField) -> None:
+        if isinstance(field.transform, TimeTransform):
+            existing_time_field = self._added_time_fields.get(field.source_id)
+            if existing_time_field:
+                raise ValueError(f"Cannot add time partition field: {field.name} conflicts with {existing_time_field.name}")
+            self._added_time_fields[field.source_id] = field
+
+    def _partition_field(self, transform_key: Tuple[int, Transform[Any, Any]], name: Optional[str]) -> PartitionField:
+        if self._table.metadata.format_version == 2:
+            source_id, transform = transform_key
+            historical_fields = []
+            for spec in self._table.specs().values():
+                for field in spec.fields:
+                    historical_fields.append((field.source_id, repr(field.transform), field.name))
+
+            for field_key in historical_fields:
+                if field_key[0] == source_id and field_key[1] == repr(transform):
+                    if name is not None or field_key[2] == name:
+                        return field
+
```

Comment:
There's a bug somewhere here in the reuse logic for v2, let me fix that. 

```
def test_multiple_remove_and_add_reuses_v2(table_v2: Table) -> None:
    with table_v2.update_spec() as update:
        update.add_field("id", BucketTransform(16), "bucketed_id")
        update.add_field("event_ts", DayTransform(), "day_ts")
    with table_v2.update_spec() as update:
        update.remove_field("day_ts").remove_field("bucketed_id")
    with table_v2.update_spec() as update:
        update.add_field("id", BucketTransform(16), "bucketed_id")
```

One would expect the bucketed_id to be the only one remaining but for some reason the DayTransform is the one that gets reused and added back.

Reply:
Ah we're returning field which is just a carryover from the loop in the previous block, that's not right.